lib/plugins/threaded-signing/plugin.c

```diff
@@ -137,29 +151,66 @@ signing_worker_thread(void *user_data)
-    g_cond_wait(&self->cond, &self->mutex);
-
-    if (self->plugin_state == THREADED_SIGNING_HAS_HASH_TO_SIGN) {
-      // Copy the |hash_to_sign| to |signature_info| and start signing. In principle, it is now
-      // possible to prepare for a new |hash_to_sign|.
+    if (self->input_buffer_idx > 0) {
+      // Get the oldest hash from the input buffer
+      // Copy the hash to |signature_info| and start signing.
       assert(self->hash_size == self->signature_info->hash_size);
-      memcpy(self->signature_info->hash, self->hash_to_sign, self->hash_size);
-      self->plugin_state = THREADED_SIGNING_SIGNS_HASH;
+      assert(self->signature_info->hash);
+      memcpy(self->signature_info->hash, self->input_buffer[0], self->hash_size);
+
+      uint8_t *tmp = self->input_buffer[0];
+      int j = 0;
+      while (self->input_buffer[j] != NULL && j < MAX_BUFFER_LENGTH - 1) {
+        self->input_buffer[j] = self->input_buffer[j + 1];
+        j++;
+      }
+      if (j < MAX_BUFFER_LENGTH - 1) {
+        self->input_buffer[j - 1] = tmp;
+      } else {
+        self->input_buffer[j] = tmp;
+      }
+      self->input_buffer_idx -= 1;
 
       // Let the signing operate outside a lock. Otherwise sv_interface_get_signature() is blocked,
       // since variables need to be read under a lock.
       g_mutex_unlock(&self->mutex);
       SignedVideoReturnCode status = openssl_sign_hash(self->signature_info);
-
+      // TODO: Remove sleeps
+      a++;
+      if (a > 3) {
+        sleep(0);
+      } else {
+        sleep(5);
+      }
       g_mutex_lock(&self->mutex);
-      // When successfully done with signing, move |plugin_state| to THREADED_SIGNING_HAS_SIGNATURE,
-      // otherwise move to THREADED_SIGNING_ERROR to report the error when getting the signature.
+
+      // If not successfully done with signing, set |signing_error| to true to
+      // report the error when getting the signature.
+      self->output_buffer[self->output_buffer_idx].signing_error = (status != SV_OK);
+
       if (status == SV_OK) {
-        self->plugin_state = THREADED_SIGNING_HAS_SIGNATURE;
-      } else {
-        self->plugin_state = THREADED_SIGNING_ERROR;
+        // Allocate memory for the |signature| if necessary.
+        if (!self->output_buffer[self->output_buffer_idx].signature) {
+          self->output_buffer[self->output_buffer_idx].signature =
+              calloc(1, self->signature_info->max_signature_size);
+          if (!self->output_buffer[self->output_buffer_idx].signature) goto catch_error;
+        }
+
+        // Copy the |signature| to the output buffer
+        memcpy(self->output_buffer[self->output_buffer_idx].signature,
+            self->signature_info->signature, self->signature_info->signature_size);
+        self->output_buffer[self->output_buffer_idx].size = self->signature_info->signature_size;
+        self->output_buffer_idx++;
       }
+    catch_error:
+      if (status == SV_UNKNOWN_FAILURE) {
```

Comment:
This will not work.
If you fail on memory allocation on line 196, ```status == SV_OK```, because otherwise you would not have entered the if-statement.

The other thing is the error handling itself. Is this really what we want? You reset all buffers and continue signing. If memory allocation fails we are in big trouble and should stop the thread.